File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        result, _ = self._isBalanced(root)
        return result
    
    def _isBalanced(self, root):
        if not root:
            return True, 0
        
        result_left, height_left = self._isBalanced(root.left)
        result_right, height_right = self._isBalanced(root.right)
        
        if result_left and result_right and abs(height_left - height_right) <= 1:
            return True, max(height_left, height_right) + 1
        else:
            return False, max(height_left, height_right) + 1

File: 110-balanced-binary-tree/balanced-binary-tree.go
/*
 * @lc app=leetcode id=110 lang=golang
 *
 * [110] Balanced Binary Tree
 *
 * https://leetcode.com/problems/balanced-binary-tree/description/
 *
 * algorithms
 * Easy (40.72%)
 * Total Accepted:    311K
 * Total Submissions: 763.5K
 * Testcase Example:  '[3,9,20,null,null,15,7]'
 *
 * Given a binary tree, determine if it is height-balanced.
 *
 * For this problem, a height-balanced binary tree is defined as:
 *
 *
 * a binary tree in which the depth of the two subtrees of every node never
 * differ by more than 1.
 *
 *
 * Example 1:
 *
 * Given the following tree [3,9,20,null,null,15,7]:
 *
 *
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 *
 * Return true.
 *
 * Example 2:
 *
 * Given the following tree [1,2,2,3,3,null,null,4,4]:
 *
 *
 * ⁠      1
 * ⁠     / \
 * ⁠    2   2
 * ⁠   / \
 * ⁠  3   3
 * ⁠ / \
 * ⁠4   4
 *
 *
 * Return false.
 *
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func dfsHeight(root *TreeNode) int {
	if root == nil {
		return 0
	}

	leftHeight := dfsHeight(root.Left)
	if leftHeight == -1 {
		return -1
	}

	rightHeight := dfsHeight(root.Right)
	if rightHeight == -1 {
		return -1
	}

	if leftHeight-rightHeight > 1 || leftHeight-rightHeight < -1 {
		return -1
	}

	return max(leftHeight, rightHeight) + 1
}
func isBalanced(root *TreeNode) bool {
	return dfsHeight(root) != -1
}
